binary-transcoders/src/main/kotlin/com/kamelia/sprinkler/transcoder/binary/decoder/composer/DecodingScope.kt

```diff
@@ -0,0 +1,483 @@
+package com.kamelia.sprinkler.transcoder.binary.decoder.composer
+
+import com.kamelia.sprinkler.transcoder.binary.decoder.core.Decoder
+import com.kamelia.sprinkler.transcoder.binary.decoder.toCollection
+import com.kamelia.sprinkler.util.ExtendedCollectors
+import java.util.stream.Collector
+import java.util.stream.Collectors
+
+/**
+ * Represents a scope in which an object can be decoded. This interface is used to compose decoders. It can be used to
+ * create a decoders in a more concise way than overriding the [Decoder] interface.
+ *
+ *
+ * In the following example, both syntaxes are almost equivalent:
+ *
+ * ```
+ * class Person(val name: String, val age: Int)
+ *
+ * // the scope syntax allows to create a decoder in a more concise way
+ * val personDecoder = composedDecoder<Person> {
+ *     val name = string()
+ *     val age = int()
+ *     Person(name, age)
```

Comment:
I mean I know it's obvious, but you may want to point out that it's a sequential DSL, by opposition to a descriptive one ; as in, because string() is called before int(), it will always try to decode a String then an Int.

This can be written in the first paragraph as simply as :

"It can be used to create a decoders in a more concise way than overriding the [Decoder] interface, **and in a sequential manner**"